### PR TITLE
✨Identity Server - Add Leaderboard

### PR DESCRIPTION
**Pull request by jackreimers**
Adds the leaderboard page and tracks each time a signed in user sends a message and stores it against their email address.
Message content is not stored.

Leaderboard can be sorted by last month, year, or all time.

Related to #103 

image
**Figure: Signed out user**

image
**Figure: Signed in user**

### CODE REVIEW

File: src/Domain/LeaderboardModel.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Domain;
+
+public class LeaderboardModel
```
Comment (Hona) [inline src/Domain/LeaderboardModel.cs:3]: should be in domain/entities

File: src/Domain/LeaderboardUser.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿namespace Domain;
+
+public class LeaderboardUser
```
Comment (Hona) [inline src/Domain/LeaderboardUser.cs:3]: should be in domain/entities

File: src/Infrastructure/RulesContext.cs [hunk 3]
```diff
@@ -476,6 +479,20 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
             entity.Property(e => e.Embeddings).HasColumnName("embeddings");
         });
 
+        modelBuilder.Entity<LeaderboardModel>(entity =>
+        {
+            entity.HasKey(e => e.Id).HasName("leaderboard_pkey");
+            
+            entity.ToTable("user_stats");
+            
+            entity.HasIndex(e => e.Id, "rules_id_key").IsUnique();
```
Comment (Hona) [inline src/Infrastructure/RulesContext.cs:488]: I would rename the index as it looks copy pasted

File: src/Infrastructure/RulesContext.cs [hunk 4]
```diff
@@ -38,6 +39,8 @@ public RulesContext(DbContextOptions<RulesContext> options)
 
     public virtual DbSet<RefreshToken> RefreshTokens { get; set; } = null!;
```
Comment (Hona) [inline src/Infrastructure/RulesContext.cs:40]: Needs EF Migrations added?

File: src/WebAPI/Routes/LeaderboardRoutes.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+                            Email = group.Key.Email,
+                            LastMonth = group.Count(u => u.Date >= oneMonth),
+                            LastYear = group.Count(u => u.Date >= oneYear),
+                            AllTime = group.Count()
+                        })
+                        .AsEnumerable();
+                    
+                    return TypedResults.Ok(result);
+                }
+            )
```
Comment (Hona) [inline src/WebAPI/Routes/LeaderboardRoutes.cs:38]: Don't do logic in the presentation layer.

This is a slim CA implementation so move this to a LeaderboardService then inject & return the method here.

e.g.

var output = service.something()
return Ok(output);

File: src/WebAPI/SignalR/RulesHub.cs [hunk 6]
```diff
@@ -77,4 +104,16 @@ CancellationToken cancellationToken
 
         return _messageHandler.Handle(messageList, apiKey, gptModel, cancellationToken);
     }
+
+    private async Task AddUserStat(string email, string name)
```
Comment (Hona) [inline src/WebAPI/SignalR/RulesHub.cs:108]: Rename to TrackUserMessage or something. 'Stat' is generic and removes context

File: src/WebUI/Pages/Leaderboard.razor [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    </MudStack>
+</div>
+
+@code {
+    private List<LeaderboardUserStats> LeaderboardData { get; set; } = null!;
+    private LeaderboardFilter Filter { get; set; }
+
+    protected override async Task OnInitializedAsync()
+    {
+        var result = await Http.GetFromJsonAsync<List<LeaderboardUserStats>>($"{Configuration.GetValue<string>("ApiBaseUrl")}/getLeaderboardStats");
```
Comment (Hona) [inline src/WebUI/Pages/Leaderboard.razor:86]: You don't need to raw dog this HTTP call - use the RulesGptApiClients that are autogenerated 

i.e. RulesGptClient.GetLeaderboardStatsAsync

**Review by Hona — CHANGES_REQUESTED** [review]

**Review by Hona — APPROVED** [review]
I'm still fairly certain this will break without a migration but lets go to staging & fix before prod

**Comment by jackreimers** [comment]
@Hona confirmed working in staging https://ashy-meadow-0a2bad900.3.azurestaticapps.net/leaderboard

**Pull request merged**